code/globalincs/undosys.h

```diff
@@ -0,0 +1,264 @@
+	}
+
+	Undo_controls.save(stack);	// Save the stack as a single item!
+	std::pair<const void*, const void*> ref = Undo_controls.undo();
+	// Undo rolls through the items as they were saved, should your data operate on the same location, this will apply the changes in the correct sequence
+	ref.second == Axis_map_to;
+	ref.first == &Axis_map_to[0];     // ::first references the _last_ item that the undo stack changed
+
+	//Also, performing an undo reverses the stack. So when you do a system undo, the changes are applied in reverse
+	ref = Undo_controls.redo();
+	ref.second == Axis_map_to;
+	ref.first == &Axis_map_to[JOY_AXIS_ACTIONS - 1];      // ::first references the _last_ item that the undo stack changed
+
+***********************************************************/
+
+/*!
+ * @brief Base class of Undo_item. Use this when making an undo stack!
+ * @sa Undo_item
+ */
+class Undo_item_base
+{
+public:
+	virtual ~Undo_item_base() = 0;
+	virtual std::pair<const void*, const void*> restore() = 0;
+};
+
+
+/*!
+ * @brief Class which handles undo operations.
+ * @details This class works with a single data item, to make a undo or redo stack you'd use something like a
+ *   std::vector or a std::deque of this class. Undo and Redo stacks would be seperate. Before restoring data from
+ *   either, you'd save a copy of the item in the opposite stack.
+ *
+ *   For example, if you're about to restore from the undo stack, you'd first save a copy of the current data to the
+ *   redo stack, then do the undo. The Undo_item in the undo stack can then be popped off the stack.
+ *
+ * @note If you want to save changes to an entire C-style array (such as 'int array[]'), try wrapping it in a std::array first.
+ *
+ * @sa Undo_item_base
+ */
+template<typename T> class Undo_item : public Undo_item_base
+{
+public:
+	Undo_item()
+		: dest(nullptr), data(nullptr) {};
+
+	Undo_item(T& _data) {
+		save(_data);
+	};
+
+	Undo_item(T& _data, T* _cont) {
+		save(_data, _cont);
+	};
+
+	~Undo_item() {
+		if (data != nullptr) {
+			delete data;
+		}
+	};
+
+	/**
+	 * @brief Restores the saved data
+	 *
+	 * @returns A pair of const void* referencing the affected item and its container (if it was provided)
+	 * @details If successful, the data at dest has been swapped. So if this was an undo item, it's now a redo item
+	 */
+	std::pair<const void*, const void*> restore() {
+		Assert((dest != nullptr) && (data != nullptr));
+
+		T copy = *static_cast<T*>(dest);
+		*static_cast<T*>(dest) = *static_cast<T*>(data);
+		*static_cast<T*>(data) = copy;
+
+		return std::pair<const void*, const void*>(dest, container);
+	};
+
+	/*!
+	 * @brief Saves data and its destination as an undo item
+	 *
+	 * @param[in] _data The data to save
+	 * @param[in] _cont The container this data is originally found in
+	 * @details The explicit definition is here to ensure type safety
+	 */
+	void save(T& _data, T* _cont = nullptr) {
```

Comment:
`_data` is a reference but you use it like a pointer (e.g. you test for `nullptr`). Is this intentional and if yes then the class should ensure that `T` is a pointer type by using a `static_assert`.

Reply:
The Assert test is something of a carry over from the first draft of this function where _data was a *.  Since its a &, testing against nullptr is moot, and shall be removed